[Docs] Add performance best practices guide with benchmarks (#10303)
pimcore/pimcore: doc/Development_Documentation/23_Installation_and_Upgrade/03_System_Setup_and_Hosting/09_Performance_Guide.md

# Performance Best-Practice Guide

> When developing a high traffic website, it is a common practice to focus on performance measures and incorporate changes to make the website even more performant. A highly optimized website drive more traffic and increase the customer base.

Reviewer: ```suggestion
When developing a high traffic website, it is a common practice to focus on performance measures and incorporate changes to make the website even more performant. A highly optimized website drives more traffic and increases the customer base.
```

To optimize the website performance, we recommend below tools and configuration. The load test performance benchmarks are noted in each section, based on [ApacheBench](https://httpd.apache.org/docs/2.4/programs/ab.html) commandline tool, for insights on comparing the average loading time of a page, before and after applying these changes.

Before we start with the load testing, it is important to turn off dev & debug mode by setting environment variables `APP_ENV=prod` & `PIMCORE_DEV_MODE=false` so as to replicate production environment scenario.

> ### PHP8 Opcache & JIT compiler

Reviewer: ```suggestion
### PHP 8 Opcache & JIT compiler
```

> Pimcore X or higher version, requires php version >= 8.0, that means we can use and easily take the benefits of newly introduced features in PHP. 

Reviewer: ```suggestion
Pimcore X or higher version, requires PHP version >= 8.0, that means we can use and easily take the benefits of newly introduced features in PHP. 
```

As you know, PHP is an interpreted language, which means the interpreter parses, compiles, and executes the code (opcode) everytime, when executing PHP scripts. This may result in wastage of CPU resources and execution time. This is where the OPcache extension comes in to play:

“OPcache improves PHP performance by storing precompiled script bytecode in shared memory, thereby removing the need for PHP to load and parse scripts on each request.”

PHP 8 has introduced the support for Just-In-Time Compilation, which has a great potential to speed up the performance. The JIT compiler requires Opcache to be enabled to work.

> By default, JIT compiler is turned off in PHP 8 and to enable it, add these configuration to ini configuration:

Reviewer: ```suggestion
By default, JIT compiler is turned off in PHP 8. To enable it, add these configuration to ini configuration:
```

```ini
opcache.enable=1
opcache.enable_cli=1
opcache.jit_buffer_size=256M
```

> `opcache.enable`: enables the Opcache for Php server

Reviewer: ```suggestion
`opcache.enable`: enables the Opcache for PHP server
```

`opcache.enable_cli`: enables the Opcache for CLI mode
`opcache.jit_buffer_size`: JIT buffer memory size. 0 value disables the JIT compiler.

#### Benchmarks:
On running command, `ab -n 100 -c 20 http://localhost/en`

- Before Changes
```bash
Time taken for tests:   1.061 seconds
Time per request:       212.203 [ms] (mean)
Time per request:       10.610 [ms] (mean, across all concurrent requests)

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    0   0.3      0       1
Processing:    65  182  41.2    205     225
Waiting:       64  182  41.3    205     224
Total:         66  183  41.0    205     225
```

- After Changes
```bash
Time taken for tests:   1.008 seconds
Time per request:       201.617 [ms] (mean)
Time per request:       10.081 [ms] (mean, across all concurrent requests)

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    0   0.4      0       1
Processing:    65  175  36.9    181     241
Waiting:       65  174  36.9    180     240
Total:         66  175  36.7    181     241
```

> ### Symfony Performance Best Practice

Reviewer: ```suggestion
### Symfony Performance Best Practices
```

As Pimcore is based on Symfony framework for backend requests processing, it is highly recommended to comply with [Symfony Performance Best Practices](https://symfony.com/doc/current/performance.html).

> ###Benchmarks:

Reviewer: ```suggestion
### Benchmarks:
```

- Before Changes
```bash
Time taken for tests:   1.100 seconds
Time per request:       219.904 [ms] (mean)
Time per request:       10.995 [ms] (mean, across all concurrent requests)

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    0   0.4      0       2
Processing:    70  194  44.6    200     289
Waiting:       69  193  44.6    200     289
Total:         70  194  44.4    200     290
```

 - After Changes
```bash
Time taken for tests:   1.042 seconds
Time per request:       208.337 [ms] (mean)
Time per request:       10.417 [ms] (mean, across all concurrent requests)

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    0   0.4      0       1
Processing:    56  182  37.1    189     254
Waiting:       55  182  37.1    189     254
Total:         56  182  36.9    189     255
```

### Opcache Preloading
Opcache preloading was introduced in PHP 7.4, a feature that could improve the performance of your application significantly, by preloading commonly used framework files and keeping the generated byte-code in memory.

In order to enable Opcache preloading, configure these settings in your ini configuration:

> ```bash

Reviewer: ```suggestion
```ini
```

opcache.preload=/var/www/home/config/preload.php

; required for opcache.preload:
opcache.preload_user=www-data
```

Note: `memory_limit` should be set to at least `200M` when using preloading with Pimcore.

It is also possible to mark which classes should be preloaded or not by using Symfony service tags [container.preload](https://symfony.com/doc/current/reference/dic_tags.html#dic-tags-container-preload) & [container.no_preload](https://symfony.com/doc/current/reference/dic_tags.html#dic-tags-container-nopreload).

#### Benchmarks:
On running command, `ab -n 100 -c 20 http://localhost/en`

- Before Changes
```bash
```

- After Changes
```bash
```

### MySQL/MariaDB Optimizations
Pimcore uses Data Access Object or DAO pattern to isolate the business logic from low level data access operations. These data access operations play a vital role in performance of Pimcore application features requiring read & write to/from the database.
Therefore, it is important to tune your database configuration to be performant and to exactly fit the needs of your application.

> #### Mysql composite index

Reviewer: ```suggestion
#### MySQL Composite Index
```

If you are using custom database tables for data persistence, then it is beneficial to optimize these tables to have composite indices, so that query optimizer uses the composite index for queries that matches all columns in the index rather than searching the complete table.
 

> #### Mysql Buffer Pool

Reviewer: ```suggestion
#### MySQL Buffer Pool
```

> The buffer pool is an area in main memory where InnoDB caches table and index data as it is accessed. The main purpose of buffer pool is to the performance by improving the response time of data retrieval. The server system variable `innodb_buffer_pool_size` can be set from 70-80% of the total available memory on a dedicated database server with only or primarily InnoDB tables

Reviewer: ```suggestion
The buffer pool is an area in main memory where InnoDB caches table and index data as it is accessed. The main purpose of buffer pool is to the performance by improving the response time of data retrieval. The server system variable `innodb_buffer_pool_size` can be set from 70-80% of the total available memory on a dedicated database server with only or primarily InnoDB tables.
```

> Please follow this [discussion](https://dba.stackexchange.com/questions/27328/how-large-should-be-mysql-innodb-buffer-pool-size) to decide buffer pool size and update my.cnf:

Reviewer: ```suggestion
Please follow this [discussion](https://dba.stackexchange.com/questions/27328/how-large-should-be-mysql-innodb-buffer-pool-size) to decide buffer pool size and update `my.cnf`:
```

> ```

Reviewer: ```suggestion
```ini
```

[mysqld]
    innodb_buffer_pool_size=5G
```

> #### Mysql Query Cache

Reviewer: ```suggestion
#### MySQL Query Cache
```

> Mysql server features a Query Cache. When enabled, the query cache stores SELECT statements together with the retrieved record set in memory, then if another identical query is received, the server can then retrieve the results from the query cache rather than parsing and executing the same query again.

Reviewer: ```suggestion
MySQL server features a Query Cache. When enabled, the query cache stores SELECT statements together with the retrieved record set in memory, then if another identical query is received, the server can then retrieve the results from the query cache rather than parsing and executing the same query again.
```

> you can check, if query cache is enabled with command `SHOW VARIABLES LIKE 'have_query_cache';`. In order to enable the query caching, update my.cnf:

Reviewer: ```suggestion
- you can check if query cache is enabled with command `SHOW VARIABLES LIKE 'have_query_cache';`. In order to enable the query caching, update `my.cnf`:
```

> ```

Reviewer: ```suggestion
```ini
```

[mysqld] 
query_cache_type=1 
query_cache_size = 10M 
query_cache_limit=256K
```

- you can also use Mysql Tuning script for more optimizations [mysqltuner.pl](https://github.com/rackerhacker/MySQLTuner-perl)

#### Benchmarks:
On running command, `ab -n 100 -c 20 http://localhost/en`

- Before Changes
```bash
```

- After Changes
```bash
```

### Pimcore Caching (Redis)

> Pimcore uses extensively caches for differently types of data. The primary cache is a pure object cache where every element (document, asset, object) in Pimcore is cached as it is (serialized objects).

Reviewer: ```suggestion
Pimcore uses extensively caches for different types of data. The primary cache is a pure object cache where every element (document, asset, object) in Pimcore is cached as it is (serialized objects).
```

Waiting:       45  120  27.7    122     165
Total:         46  120  27.5    123     166
```

- After Changes
```bash
Time taken for tests:   0.646 seconds
Time per request:       129.298 [ms] (mean)
Time per request:       6.465 [ms] (mean, across all concurrent requests)

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    0   0.3      0       1
Processing:    45  115  26.0    116     167
Waiting:       45  114  26.0    116     166
Total:         46  115  25.9    116     167
```

TODO: add session topic
- Only use sessions when really necessary. The Pimcore full-page cache detects the usage of sessions in the code and disables itself if necessary.

### Static-Page-Generator
Pimcore offers a static page generator service, which generates static HTML files out of Twig templates processed in MVC framework. These static files can be delivered directly instead of going through complete MVC cycle on a frontend request. This feature is not recommended for a page with dynamic content i.e. news listing, product pages, and so on, where content changes frequently.

#### Enable Static Page generator for a Document:

> To enable automatic static page generation for a document, go to Document -> Settings -> Satic Page Generator. Mark enable checkbox and define optional lifetime for static pages (which regenerates static page after lifetime) and save document.

Reviewer: ```suggestion
To enable automatic static page generation for a document, go to Document -> Settings -> Static Page Generator. Mark enable checkbox and define optional lifetime for static pages (which regenerates static page after lifetime) and save document.
```
